### PR TITLE
Feature/ret 1654 write message about breaking test

### PR DESCRIPTION
**Pull request by ivonx**

### CODE REVIEW

**Comment by beatngu13** [comment]
Could you mark this as a draft PR?

File: src/main/java/de/retest/web/RecheckSeleniumAdapter.java [hunk 1]
```diff
@@ -82,6 +83,7 @@ public boolean canCheck( final Object toVerify ) {
 
 		if ( driver instanceof UnbreakableDriver ) {
 			((UnbreakableDriver) driver).setLastActualState( lastChecked );
+			this.driver = (UnbreakableDriver) driver;
```
Comment (beatngu13) [inline src/main/java/de/retest/web/RecheckSeleniumAdapter.java]: You dont have to do this cast twice:

```java
if ( driver instanceof UnbreakableDriver ) {
    this.driver = (UnbreakableDriver) driver;
    this.driver.setLastActualState( lastChecked );
}
```

File: src/main/java/de/retest/web/RecheckSeleniumAdapter.java [hunk 2]
```diff
@@ -47,6 +46,8 @@
 
 	private final DefaultValueFinder defaultValueFinder = new DefaultWebValueFinder();
 
+	private UnbreakableDriver driver;
```
Comment (beatngu13) [inline src/main/java/de/retest/web/RecheckSeleniumAdapter.java]: What do we have this field for? Is it used somewhere?

File: src/main/java/de/retest/web/selenium/TestHealer.java [hunk 3]
```diff
@@ -226,4 +234,18 @@ private void writeWarnLogForChangedIdentifier( final String elementIdentifier, f
 		}
 	}
 
+	public static void addWarning( final String retestId, final String elementIdentifier, final Warning warning ) {
```
Comment (beatngu13) [inline src/main/java/de/retest/web/selenium/TestHealer.java]: Where is this going to be called?

File: src/main/java/de/retest/web/selenium/TestHealer.java [hunk 4]
```diff
@@ -226,4 +234,18 @@ private void writeWarnLogForChangedIdentifier( final String elementIdentifier, f
...
+		for ( final ElementDifference elementDifference : stateDifference.getElementDifferences() ) {
+			if ( elementDifference.getRetestId().equals( retestId ) ) {
+				final AttributesDifference attributesDifference = elementDifference.getAttributesDifference();
+				for ( final AttributeDifference attributeDifference : attributesDifference.getDifferences() ) {
+					if ( attributeDifference.getKey().equals( elementIdentifier ) ) {
+						attributeDifference.setWarning( warning );
+					}
+				}
+			}
+
```
Comment (beatngu13) [inline src/main/java/de/retest/web/selenium/TestHealer.java]: This linebreak can be removed.

File: src/main/java/de/retest/web/selenium/UnbreakableDriver.java [hunk 5]
```diff
@@ -56,6 +58,14 @@ public RootElement getLastExpectedState() {
 		return lastExpectedState;
 	}
 
+	public ActionReplayResult getLastActionReplayResult() {
```
Comment (beatngu13) [inline src/main/java/de/retest/web/selenium/UnbreakableDriver.java]: `UnbreakableDriver` is public API, so I assume we should add Javadoc here?

File: src/main/java/de/retest/web/selenium/TestHealer.java [hunk 6]
```diff
@@ -35,6 +43,7 @@
 	private final UnbreakableDriver wrapped;
 	private final RootElement lastExpectedState;
 	private final RootElement lastActualState;
+	private static ActionReplayResult lastActionReplayResult;
```
Comment (beatngu13) [inline src/main/java/de/retest/web/selenium/TestHealer.java]: Why a _static_ mutable field? In [line 55](https://github.com/retest/recheck-web/pull/251/files#diff-2b1201bc248da98862b85ba88ea22564R55), it's not a static invocation.

File: src/main/java/de/retest/web/selenium/TestHealer.java [hunk 7]
```diff
@@ -224,6 +238,50 @@ private void writeWarnLogForChangedIdentifier( final String elementIdentifier, f
 		} else {
 			logger.warn( "Use `By.retestId(\"{}\")` to update your test {}.", retestId, callLocation );
 		}
+		addWarning( elementIdentifier, byMethodName, oldValue, newValue, test, callLocation, retestId );
+	}
+
+	public static void addWarning( final String elementIdentifier, final String byMethodName, final Object oldValue,
```
Comment (beatngu13) [inline src/main/java/de/retest/web/selenium/TestHealer.java]: Why is this (and the nested methods) static and public?

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by rebazer** [comment]
Rebase failed, this pull request needs some manual love ...

**Comment by beatngu13** [comment]
Simply reopen when rebase is done…

**Pull request closed without merging**